File: engine/skills/lib/mailbox_io.py
#!/usr/bin/env python3
"""Shared helpers for append-only mailbox reads and writes."""

from __future__ import annotations

import json
import os
from datetime import datetime, timezone
from pathlib import Path

import fcntl


def utcnow() -> str:
    return datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")


def _valid_message_id(value: object) -> str | None:
    if not isinstance(value, str):
        return None
    if not value.startswith("mail."):
        return None
    parts = value.split(".")
    if len(parts) != 3:
        return None
    if not parts[1].endswith("Z"):
        return None
    if not parts[2].isdigit():
        return None
    return value


def _next_message_id(lines: list[str], timestamp: str) -> str:
    stamp = datetime.now(timezone.utc).strftime("%Y%m%dT%H%M%SZ")
    prefix = f"mail.{stamp}."
    seq = 1

    for raw_line in reversed(lines):
        line = raw_line.strip()
        if not line:
            continue
        try:
            data = json.loads(line)
        except json.JSONDecodeError:
            continue

        existing_id = _valid_message_id(data.get("id"))
        if not existing_id or not existing_id.startswith(prefix):
            if existing_id:
                break
            continue

        seq = int(existing_id.rsplit(".", 1)[-1]) + 1
        break

    return f"{prefix}{seq:03d}"


def read_messages(mailbox_path: Path) -> list[dict]:
    if not mailbox_path.exists():
        return []

    messages: list[dict] = []
    for raw_line in mailbox_path.read_text(encoding="utf-8").splitlines():
        line = raw_line.strip()
        if not line:
            continue
        try:
            data = json.loads(line)
        except json.JSONDecodeError:
            continue
        if isinstance(data, dict):
            messages.append(data)
    return messages


def messages_after_id(
    mailbox_path: Path,
    last_id: str | None,
    sender: str | None = None,
) -> list[dict]:
    messages = read_messages(mailbox_path)
    if last_id is None:
        return [msg for msg in messages if sender is None or msg.get("from") == sender]

    pending: list[dict] = []
    found = False

    for message in messages:
        if not found:
            if message.get("id") == last_id:
                found = True
            continue
        if sender is None or message.get("from") == sender:
            pending.append(message)

    if found:
        return pending

    return [msg for msg in messages if sender is None or msg.get("from") == sender]


def mailbox_contains_field(mailbox_path: Path, field: str, expected_value: str) -> bool:
    for message in reversed(read_messages(mailbox_path)):
        if message.get(field) == expected_value:
            return True
    return False


def append_message(
    mailbox_path: Path,
    from_id: str,
    to_id: str,
    task_id: str,
    message: str,
    extra: dict | None = None,
) -> dict:
    mailbox_path.parent.mkdir(parents=True, exist_ok=True)

    clean_message = message.strip()
    if not clean_message:
        raise ValueError("message must not be empty")

    timestamp = utcnow()

    with mailbox_path.open("a+", encoding="utf-8") as handle:
        fcntl.flock(handle.fileno(), fcntl.LOCK_EX)
        handle.seek(0)
        lines = handle.read().splitlines()

        entry = {
            "id": _next_message_id(lines, timestamp),
            "ts": timestamp,
            "from": from_id,
            "to": to_id,
            "task_id": task_id,
            "message": clean_message,
        }
        if extra:
            entry.update(extra)

        handle.seek(0, os.SEEK_END)
        handle.write(json.dumps(entry, ensure_ascii=False) + "\n")
        handle.flush()
        os.fsync(handle.fileno())
        fcntl.flock(handle.fileno(), fcntl.LOCK_UN)

    return entry


def load_contacts(contacts_path: Path) -> dict:
    if not contacts_path.exists():
        return {}
    try:
        return json.loads(contacts_path.read_text(encoding="utf-8"))
    except (json.JSONDecodeError, OSError):
        return {}


def save_contacts(contacts_path: Path, contacts: dict) -> None:
    contacts_path.parent.mkdir(parents=True, exist_ok=True)
    tmp = contacts_path.with_suffix(".tmp")
    tmp.write_text(json.dumps(contacts, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")
    tmp.rename(contacts_path)


def write_pending_message(runtime_dir: Path, contact: str, mailbox_id: str, source: str) -> None:
    pending_dir = runtime_dir / "pending_messages"
    pending_dir.mkdir(parents=True, exist_ok=True)
    pending_file = pending_dir / f"{contact}.json"
    data = {"mailbox_id": mailbox_id, "ts": utcnow(), "source": source}
    pending_file.write_text(json.dumps(data, ensure_ascii=False) + "\n", encoding="utf-8")


def load_pending_messages(runtime_dir: Path) -> dict[str, dict]:
    pending_dir = runtime_dir / "pending_messages"
    if not pending_dir.is_dir():
        return {}
    result = {}
    for f in pending_dir.glob("*.json"):
        try:
            data = json.loads(f.read_text(encoding="utf-8"))
            if isinstance(data, dict):
                contact = f.stem
                result[contact] = data
        except (json.JSONDecodeError, OSError):
            continue
    return result


def clear_pending_message(runtime_dir: Path, contact: str, expected_mailbox_id: str | None) -> None:
    if not expected_mailbox_id:
        return
    pending_file = runtime_dir / "pending_messages" / f"{contact}.json"
    if not pending_file.exists():
        return
    try:
        data = json.loads(pending_file.read_text(encoding="utf-8"))
        if data.get("mailbox_id") == expected_mailbox_id:
            pending_file.unlink(missing_ok=True)
    except (json.JSONDecodeError, OSError):
        pass


def count_unread(mailbox_path: Path, last_id: str | None) -> int:
    return len(messages_after_id(mailbox_path, last_id))

File: engine/src/shared/mailbox-io.ts
import fs from "node:fs";
import path from "node:path";
import { utcnow } from "../harness-core/time.js";

export interface MailboxEntry {
  id: string;
  ts: string;
  from: string;
  to: string;
  task_id: string;
  message: string;
  [k: string]: unknown;
}

export function readMessages(mailboxPath: string): MailboxEntry[] {
  if (!fs.existsSync(mailboxPath)) return [];
  const text = fs.readFileSync(mailboxPath, "utf8");
  const out: MailboxEntry[] = [];
  for (const line of text.split("\n")) {
    const trimmed = line.trim();
    if (!trimmed) continue;
    try {
      const data = JSON.parse(trimmed);
      if (data && typeof data === "object") out.push(data as MailboxEntry);
    } catch {
      /* skip malformed */
    }
  }
  return out;
}

export function loadContacts(contactsPath: string): Record<string, unknown> {
  if (!fs.existsSync(contactsPath)) return {};
  try {
    const data = JSON.parse(fs.readFileSync(contactsPath, "utf8"));
    return typeof data === "object" && data !== null ? (data as Record<string, unknown>) : {};
  } catch {
    return {};
  }
}

export function saveContacts(contactsPath: string, contacts: Record<string, unknown>): void {
  fs.mkdirSync(path.dirname(contactsPath), { recursive: true });
  const tmp = contactsPath + ".tmp";
  fs.writeFileSync(tmp, JSON.stringify(contacts, null, 2) + "\n", "utf8");
  fs.renameSync(tmp, contactsPath);
}

function validMessageId(value: unknown): string | null {
  if (typeof value !== "string") return null;
  if (!value.startsWith("mail.")) return null;
  const parts = value.split(".");
  if (parts.length !== 3) return null;
  if (!parts[1].endsWith("Z")) return null;
  if (!/^\d+$/.test(parts[2])) return null;
  return value;
}

function nextMessageId(lines: string[]): string {
  const stamp = new Date().toISOString().replace(/[-:]/g, "").replace(/\.\d{3}Z$/, "Z");
  const prefix = `mail.${stamp}.`;
  let seq = 1;
  for (let i = lines.length - 1; i >= 0; i--) {
    const line = lines[i].trim();
    if (!line) continue;
    let data: { id?: unknown };
    try {
      data = JSON.parse(line);
    } catch {
      continue;
    }
    const id = validMessageId(data.id);
    if (!id) continue;
    if (!id.startsWith(prefix)) break;
    seq = parseInt(id.split(".").pop()!, 10) + 1;
    break;
  }
  return `${prefix}${String(seq).padStart(3, "0")}`;
}

export function appendMessage(
  mailboxPath: string,
  fromId: string,
  toId: string,
  taskId: string,
  message: string,
  extra?: Record<string, unknown>
): MailboxEntry {
  fs.mkdirSync(path.dirname(mailboxPath), { recursive: true });
  const clean = message.trim();
  if (!clean) throw new Error("message must not be empty");

  const existing = fs.existsSync(mailboxPath) ? fs.readFileSync(mailboxPath, "utf8").split("\n") : [];
  const entry: MailboxEntry = {
    id: nextMessageId(existing),
    ts: utcnow(),
    from: fromId,
    to: toId,
    task_id: taskId,
    message: clean,
    ...(extra ?? {}),
  };
  fs.appendFileSync(mailboxPath, JSON.stringify(entry) + "\n", "utf8");
  return entry;
}

export function messagesAfterId(
  mailboxPath: string,
  lastId: string | null,
  sender?: string | null
): MailboxEntry[] {
  const messages = readMessages(mailboxPath);
  const match = (m: MailboxEntry) => sender == null || m.from === sender;
  if (!lastId) return messages.filter(match);

  const out: MailboxEntry[] = [];
  let found = false;
  for (const m of messages) {
    if (!found) {
      if (m.id === lastId) found = true;
      continue;
    }
    if (match(m)) out.push(m);
  }
  return found ? out : messages.filter(match);
}

export function mailboxContainsField(
  mailboxPath: string,
  field: string,
  expectedValue: string
): boolean {
  const messages = readMessages(mailboxPath);
  for (let i = messages.length - 1; i >= 0; i--) {
    if (messages[i][field] === expectedValue) return true;
  }
  return false;
}

export function writePendingMessage(
  runtimeDir: string,
  contact: string,
  mailboxId: string,
  source: string
): void {
  const pendingDir = path.join(runtimeDir, "pending_messages");
  fs.mkdirSync(pendingDir, { recursive: true });
  const file = path.join(pendingDir, `${contact}.json`);
  fs.writeFileSync(
    file,
    JSON.stringify({ mailbox_id: mailboxId, ts: utcnow(), source }) + "\n",
    "utf8"
  );
}
